### PR TITLE
Bulk add products to taxon

### PR DESCRIPTION
**Pull request by igormukhingmailcom**
And some improvements at existing bulk edit action...

image

### CODE REVIEW

File: src/Controller/Action/BulkAddTaxonsToProductsAction.php [hunk 1]
```diff
@@ -0,0 +1,98 @@
...
+                    $productTaxon = $this->productTaxonFactory->createNew();
+                    $productTaxon->setTaxon($taxon);
+
+                    $product->addProductTaxon($productTaxon);
+                }
+            }
+
+            $this->productManager->flush();
+
+            $redirectUrl = $request->getUri();
```
Comment (loevgaard) [inline src/Controller/Action/BulkAddTaxonsToProductsAction.php]: Redundant

File: src/Controller/Action/BulkAddTaxonsToProductsAction.php [hunk 2]
```diff
@@ -0,0 +1,98 @@
...
+
+            return new RedirectResponse($redirectUrl);
+        }
+
+        $editAction = $this->router->generate('setono_sylius_bulk_edit_admin_bulk_edit_products');
```
Comment (loevgaard) [inline src/Controller/Action/BulkAddTaxonsToProductsAction.php]: Why is this needed?

**Review by loevgaard — COMMENTED** [review]
Very nice PR. Can we rename the templates with underscore instead of camel case?

I am on my mobile so need to check this PR later also :)

File: src/Controller/Action/BulkEditProductsAction.php [hunk 3]
```diff
@@ -25,27 +27,41 @@ final class BulkEditProductsAction
...
 
+    private RouterInterface $router;
+
     public function __construct(
         ProductRepositoryInterface $productRepository,
         ProductVariantRepositoryInterface $productVariantRepository,
         ChannelRepositoryInterface $channelRepository,
-        Environment $twig
+        Environment $twig,
+        RouterInterface $router
```
Comment (loevgaard) [inline src/Controller/Action/BulkEditProductsAction.php]: I think the `UrlGeneratorInterface` is enough here

File: src/Controller/Action/BulkEditProductsAction.php [hunk 4]
```diff
@@ -25,27 +27,41 @@ final class BulkEditProductsAction
...
         /** @var ChannelInterface[] $channels */
         $channels = $this->channelRepository->findAll();
+        Assert::minCount($channels, 1);
 
         if (!$request->query->has('channelCode')) {
-            return new RedirectResponse($request->getUri() . '&channelCode=' . $channels[0]->getCode());
+            $channelAwareUrl = $request->getUri();
+            $channelAwareUrl .= '&channelCode=' . $channels[0]->getCode();
+
+            return new RedirectResponse($channelAwareUrl);
```
Comment (loevgaard) [inline src/Controller/Action/BulkEditProductsAction.php:56]: Why was this oneliner converted into four lines?

File: src/Controller/Action/BulkAddTaxonsToProductsAction.php (hunk 2 above)
Comment (loevgaard) [reply src/Controller/Action/BulkAddTaxonsToProductsAction.php]: Ah, this is needed to create a link from the bulk taxon to the bulk edit and vice versa?

File: src/Resources/translations/messages.da.yaml [hunk 5]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
     ui:
         bulk_edit: Masseredigering
         bulk_edit_products: Masseredigering af produkter
+        bulk_add_taxons: Bulk add taxons
+        bulk_add_products_to_taxons: Bulk add taxons
+        present_taxons: Present taxons
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]: Should this be 'existing taxons' instead?

File: src/Resources/translations/messages.da.yaml [hunk 6]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
...
+        selected_products: '%products_count% selected products'
+        action:
+            add_taxons: Add taxons
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]:
```suggestion
            add_taxons: Tilføj taksonomier
```

File: src/Resources/translations/messages.da.yaml [hunk 7]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
...
+            add_taxons_selected: Add taxons to selected
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]:
```suggestion
            add_taxons_selected: Tilføj taksonomier til valgte
```

File: src/Resources/translations/messages.da.yaml [hunk 8]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
...
+            edit: Rediger
+            edit_selected: Edit selected
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]:
```suggestion
            edit_selected: Rediger valgte
```

File: src/Resources/translations/messages.da.yaml [hunk 9]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
     ui:
         bulk_edit: Masseredigering
         bulk_edit_products: Masseredigering af produkter
+        bulk_add_taxons: Bulk add taxons
+        bulk_add_products_to_taxons: Bulk add taxons
+        present_taxons: Present taxons
+        selected_products: '%products_count% selected products'
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]:
```suggestion
        selected_products: '%products_count% valgte produkter'
```

File: src/Resources/translations/messages.da.yaml [hunk 10]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
     ui:
         bulk_edit: Masseredigering
         bulk_edit_products: Masseredigering af produkter
+        bulk_add_taxons: Bulk add taxons
+        bulk_add_products_to_taxons: Bulk add taxons
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]:
```suggestion
        bulk_add_products_to_taxons: Bulk tilføj taksonomier
```

File: src/Resources/translations/messages.da.yaml [hunk 11]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
     ui:
         bulk_edit: Masseredigering
         bulk_edit_products: Masseredigering af produkter
+        bulk_add_taxons: Bulk add taxons
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]:
```suggestion
        bulk_add_taxons: Bulk tilføj taksonomier
```

File: src/Resources/translations/messages.da.yaml [hunk 12]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
...
+        bulk_add_products_to_taxons: Bulk add taxons
+        present_taxons: Present taxons
+        selected_products: '%products_count% selected products'
+        action:
+            add_taxons: Add taxons
+            add_taxons_selected: Add taxons to selected
+            edit: Rediger
+            edit_selected: Edit selected
+    form:
+        taxons: Select taxons to add products to
```
Comment (loevgaard) [inline src/Resources/translations/messages.da.yaml]:
```suggestion
        taxons: Vælg taksonomier at tilføje til produkter
```

File: src/Resources/translations/messages.en.yaml [hunk 13]
```diff
@@ -2,3 +2,14 @@ setono_sylius_bulk_edit:
...
+        bulk_add_products_to_taxons: Bulk add taxons
+        present_taxons: Present taxons
+        selected_products: '%products_count% selected products'
+        action:
+            add_taxons: Add taxons
+            add_taxons_selected: Add taxons to selected
+            edit: Edit
+            edit_selected: Edit selected
+    form:
+        taxons: Select taxons to add products to
```
Comment (loevgaard) [inline src/Resources/translations/messages.en.yaml]:
```suggestion
        taxons: Select taxons to add to products
```

**Review by loevgaard — CHANGES_REQUESTED** [review]

File: src/Controller/Action/BulkAddTaxonsToProductsAction.php (hunk 2 above)
Comment (igormukhingmailcom) [reply src/Controller/Action/BulkAddTaxonsToProductsAction.php]: Yeah

**Comment by igormukhingmailcom** [comment]
> Can we rename the templates with underscore instead of camel case?

If you mean `_headerTitle.html.twig` - I was trying to keep structure like original Sylius Admin Crud templates have.

File: src/Controller/Action/BulkEditProductsAction.php (hunk 4 above)
Comment (igormukhingmailcom) [reply src/Controller/Action/BulkEditProductsAction.php:56]: To have less width :)

**Review by loevgaard — APPROVED** [review]
